### PR TITLE
Wire a Retryable property source locator when spring-retry is on the cla...

### PR DESCRIPTION
**Pull request by roykachouh**
...sspath. #72

### CODE REVIEW

File: spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/RetryableConfigServicePropertySourceLocator.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+	public void setRetryTemplate(RetryTemplate retryTemplate) {
+		this.retryTemplate = retryTemplate;
+	}
+
+	private RetryTemplate getRetryTemplate(ConfigClientProperties client) {
+		RetryTemplate retryTemplate = new RetryTemplate();
+		retryTemplate.setRetryPolicy(new NeverRetryPolicy());
+
+		if (client.getRetryBeforeFail() > 0) {
```
Comment (roykachouh) [inline spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/RetryableConfigServicePropertySourceLocator.java]: I think the formatting I got from https://github.com/spring-cloud/spring-cloud-build/blob/master/spring-cloud-build/eclipse-code-formatter.xml is either not accurate or their is something wrong with the eclipse formatter plugin for IDEA.

File: spring-cloud-config-client/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java [hunk 2]
```diff
@@ -143,12 +142,23 @@ public ConfigClientProperties configClientProperties() {
 		}
 
 		@Bean
+		@ConditionalOnMissingClass(name = {"org.springframework.cloud.config.server.EnvironmentRepository", "org.springframework.retry.RetryOperations"})
```
Comment (roykachouh) [inline spring-cloud-config-client/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: I'm not sure if this is what we had in mind for disabling the PropertySourceLocator during server bootstrap phase.

Comment (spencergibb) [inline spring-cloud-config-client/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: Why the `ConditionalOnMissingClass`?

File: spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/RetryableConfigServicePropertySourceLocator.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.retry.policy.SimpleRetryPolicy;
+import org.springframework.retry.support.RetryTemplate;
+
+public class RetryableConfigServicePropertySourceLocator extends
+		AbstractConfigServicePropertyLocator {
+
+	private static Log logger = LogFactory
+			.getLog(RetryableConfigServicePropertySourceLocator.class);
+
+	private RetryTemplate retryTemplate;
```
Comment (spencergibb) [inline spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/RetryableConfigServicePropertySourceLocator.java]: Is `retryTemplate` used?  A new one is created in `getRetryTemplate()`.

Comment (roykachouh) [inline spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/RetryableConfigServicePropertySourceLocator.java]: Good catch, I corrected that as well as the `RestTemplate` inside `ConfigServicePropertySourceLocator`

File: spring-cloud-config-client/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java (hunk 2 above)
Comment (roykachouh) [inline spring-cloud-config-client/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: Is there a better way to address what was discussed here: https://github.com/spring-cloud/spring-cloud-config/issues/72#issuecomment-75766916 ?

Sorry, if i missed something obvious, but basically I think we need to disable the `PropertySourceLocator` when the config server is bootstrapping.

Comment (spencergibb) [inline spring-cloud-config-client/src/main/java/org/springframework/cloud/bootstrap/config/PropertySourceBootstrapConfiguration.java]: I guess it won't hurt.  This is in the client module.

**Comment by dsyer** [comment]
Would it be simpler to just mark the `ConfigServicePropertySourceLocator` as `@Retryable` and then provide a `RetryTemplate` if Spring Retry is on the classpath at runtime?

**Comment by roykachouh** [comment]
Thanks Dave.  Just wanted to make sure I am on the same page as you.

I looked into this more, are you referring to enabling the `@Retryable` annotation conditional on spring-retry being on the classpath?  

Something like this?

```
@Configuration
	@ConditionalOnClass(value = RetryOperations.class)
	@EnableRetry
	protected static class PropertySourceLocatorRetryConfiguration {}
```

If we go with the above, I think we will now have to include a dependency on `org.aspectj`, and how we will allow for the `ConfigClientProperties` to dictate the retry semantics if we use the declarative `@Retryable`?

**Comment by dsyer** [comment]
You can use the `@Retryable(interceptor="beanName")` form.

**Comment by talksinmath** [comment]
I'm sorry if this question is misplaced here, but I was wondering if this is gonna make it into master anytime soon? It would help our project a lot....

Thanks!

**Comment by talksinmath** [comment]
Thanks! Awesome :-)

**Comment by talksinmath** [comment]
I have a problem with the 1.0.1 - it seems that EnableRetry does not work (at least not in the way it did in the code from @roykachouh).
Am I missing something? To me it seems that at least two classes (RetryableConfigServicePropertySourceLocator and RefreshEndpoint) are missing from the pull...

Sorry if I am missing something obvious and thanks a lot!!

**Comment by dsyer** [comment]
I didn't use the pull request at all. What are you doing and what are you expecting to happen and what is actually happening?

**Comment by talksinmath** [comment]
I tried the code from @roykachouh and it worked fine (retries are being tried until the Config Server is up and running).
I was under the impression that this functionality was merged into the 1.0.1 version. But, using the same code in my app and version 1.0.1 I do not see any retries are being made at all.

Thanks for your help!

**Comment by dsyer** [comment]
This code is not merged, but a change was made. You should see it working if you include Spring Retry and `spring-boot-starter-aop` on your classpath (and also set `spring.cloud.config.failFast=true` of course).

**Comment by talksinmath** [comment]
I have (amongst others) spring-boot-starter-aop, spring-cloud-config-client, spring-retry on the classpath. My application has EnableRetry annotated. failFast is set to true.

--> The Application starts, prints this:
 ConfigServicePropertySourceLocator: Could not locate PropertySource: I/O error on GET request for "http://localhost:8080/configserver/elasticsearch/default":Connection refused

No retries are made and the app fails.

**Comment by dsyer** [comment]
Can you post your code (or better still a minimal project that reproduces the problem)?

**Comment by talksinmath** [comment]
Before I do that (working on it), one little question:
What is the expected behaviour? In the code from @roykachouh the app was blocked while trying to contact the config server and would only resume the startup process once the configserver was available (or the retries maxxed out eventually).

Is this the expected behaviour in 1.0.1 ?

**Comment by dsyer** [comment]
Yes, that is expected. See https://github.com/spring-cloud/spring-cloud-config/blob/master/spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServiceBootstrapConfiguration.java#L68.

**Comment by talksinmath** [comment]
https://www.dropbox.com/l/ZJKEYoAz9hxXqAjpJfJyLr

That is my sample project. It starts, doesn't find the config-server and fails (since failFast=true).

Thanks again, I really appreciate your help!

**Comment by dsyer** [comment]
Your project works for me. It tries to load the config server properties 6 times and then gives up:

```
[2015-04-27 15:18:17.888] boot - 27735 DEBUG [main] --- RetryTemplate: Retry: count=0
[2015-04-27 15:18:18.259] boot - 27735 DEBUG [main] --- ExponentialBackOffPolicy: Sleeping for 1000
[2015-04-27 15:18:19.259] boot - 27735 DEBUG [main] --- RetryTemplate: Checking for rethrow: count=1
[2015-04-27 15:18:19.260] boot - 27735 DEBUG [main] --- RetryTemplate: Retry: count=1
[2015-04-27 15:18:19.279] boot - 27735 DEBUG [main] --- ExponentialBackOffPolicy: Sleeping for 1100
[2015-04-27 15:18:20.380] boot - 27735 DEBUG [main] --- RetryTemplate: Checking for rethrow: count=2
[2015-04-27 15:18:20.380] boot - 27735 DEBUG [main] --- RetryTemplate: Retry: count=2
[2015-04-27 15:18:20.406] boot - 27735 DEBUG [main] --- ExponentialBackOffPolicy: Sleeping for 1210
[2015-04-27 15:18:21.616] boot - 27735 DEBUG [main] --- RetryTemplate: Checking for rethrow: count=3
[2015-04-27 15:18:21.617] boot - 27735 DEBUG [main] --- RetryTemplate: Retry: count=3
[2015-04-27 15:18:21.641] boot - 27735 DEBUG [main] --- ExponentialBackOffPolicy: Sleeping for 1331
[2015-04-27 15:18:22.972] boot - 27735 DEBUG [main] --- RetryTemplate: Checking for rethrow: count=4
[2015-04-27 15:18:22.973] boot - 27735 DEBUG [main] --- RetryTemplate: Retry: count=4
[2015-04-27 15:18:23.002] boot - 27735 DEBUG [main] --- ExponentialBackOffPolicy: Sleeping for 1464
[2015-04-27 15:18:24.467] boot - 27735 DEBUG [main] --- RetryTemplate: Checking for rethrow: count=5
[2015-04-27 15:18:24.467] boot - 27735 DEBUG [main] --- RetryTemplate: Retry: count=5
[2015-04-27 15:18:24.484] boot - 27735 DEBUG [main] --- RetryTemplate: Checking for rethrow: count=6
[2015-04-27 15:18:24.484] boot - 27735 DEBUG [main] --- RetryTemplate: Retry failed last attempt: count=6
[2015-04-27 15:18:24.485] boot - 27735  INFO [main] --- ClasspathLoggingApplicationListener: Application failed to start with classpath: [file:/home/dsyer/dev/cloud/workspace/foo/target/classes/, file:/home/dsyer/.m2/repository/org/springframework/boot/spring-boot-starter/1.2.3.RELEASE/spring-boot-starter-1.2.3.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/boot/spring-boot/1.2.3.RELEASE/spring-boot-1.2.3.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/spring-context/4.1.6.RELEASE/spring-context-4.1.6.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/boot/spring-boot-autoconfigure/1.2.3.RELEASE/spring-boot-autoconfigure-1.2.3.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/spring-core/4.1.6.RELEASE/spring-core-4.1.6.RELEASE.jar, file:/home/dsyer/.m2/repository/org/yaml/snakeyaml/1.14/snakeyaml-1.14.jar, file:/home/dsyer/.m2/repository/org/springframework/boot/spring-boot-starter-web/1.2.3.RELEASE/spring-boot-starter-web-1.2.3.RELEASE.jar, file:/home/dsyer/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.4.5/jackson-databind-2.4.5.jar, file:/home/dsyer/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.4.5/jackson-annotations-2.4.5.jar, file:/home/dsyer/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.4.5/jackson-core-2.4.5.jar, file:/home/dsyer/.m2/repository/org/hibernate/hibernate-validator/5.1.3.Final/hibernate-validator-5.1.3.Final.jar, file:/home/dsyer/.m2/repository/org/jboss/logging/jboss-logging/3.1.3.GA/jboss-logging-3.1.3.GA.jar, file:/home/dsyer/.m2/repository/com/fasterxml/classmate/1.0.0/classmate-1.0.0.jar, file:/home/dsyer/.m2/repository/org/springframework/spring-web/4.1.6.RELEASE/spring-web-4.1.6.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/spring-beans/4.1.6.RELEASE/spring-beans-4.1.6.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/spring-webmvc/4.1.6.RELEASE/spring-webmvc-4.1.6.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/spring-expression/4.1.6.RELEASE/spring-expression-4.1.6.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/boot/spring-boot-starter-tomcat/1.2.3.RELEASE/spring-boot-starter-tomcat-1.2.3.RELEASE.jar, file:/home/dsyer/.m2/repository/org/apache/tomcat/embed/tomcat-embed-core/8.0.20/tomcat-embed-core-8.0.20.jar, file:/home/dsyer/.m2/repository/org/apache/tomcat/embed/tomcat-embed-el/8.0.20/tomcat-embed-el-8.0.20.jar, file:/home/dsyer/.m2/repository/org/apache/tomcat/embed/tomcat-embed-logging-juli/8.0.20/tomcat-embed-logging-juli-8.0.20.jar, file:/home/dsyer/.m2/repository/org/apache/tomcat/embed/tomcat-embed-websocket/8.0.20/tomcat-embed-websocket-8.0.20.jar, file:/home/dsyer/.m2/repository/org/springframework/boot/spring-boot-starter-log4j/1.2.3.RELEASE/spring-boot-starter-log4j-1.2.3.RELEASE.jar, file:/home/dsyer/.m2/repository/org/slf4j/jcl-over-slf4j/1.7.11/jcl-over-slf4j-1.7.11.jar, file:/home/dsyer/.m2/repository/org/slf4j/slf4j-api/1.7.11/slf4j-api-1.7.11.jar, file:/home/dsyer/.m2/repository/org/slf4j/jul-to-slf4j/1.7.11/jul-to-slf4j-1.7.11.jar, file:/home/dsyer/.m2/repository/org/slf4j/slf4j-log4j12/1.7.11/slf4j-log4j12-1.7.11.jar, file:/home/dsyer/.m2/repository/log4j/log4j/1.2.17/log4j-1.2.17.jar, file:/home/dsyer/.m2/repository/org/springframework/cloud/spring-cloud-starter-eureka/1.0.1.RELEASE/spring-cloud-starter-eureka-1.0.1.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/cloud/spring-cloud-starter/1.0.1.RELEASE/spring-cloud-starter-1.0.1.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/cloud/spring-cloud-netflix-core/1.0.1.RELEASE/spring-cloud-netflix-core-1.0.1.RELEASE.jar, file:/home/dsyer/.m2/repository/com/netflix/eureka/eureka-client/1.1.147/eureka-client-1.1.147.jar, file:/home/dsyer/.m2/repository/org/codehaus/jettison/jettison/1.2/jettison-1.2.jar, file:/home/dsyer/.m2/repository/stax/stax-api/1.0.1/stax-api-1.0.1.jar, file:/home/dsyer/.m2/repository/com/netflix/netflix-commons/netflix-eventbus/0.1.2/netflix-eventbus-0.1.2.jar, file:/home/dsyer/.m2/repository/com/netflix/archaius/archaius-core/0.6.5/archaius-core-0.6.5.jar, file:/home/dsyer/.m2/repository/com/google/guava/guava/18.0/guava-18.0.jar, file:/home/dsyer/.m2/repository/commons-configuration/commons-configuration/1.8/commons-configuration-1.8.jar, file:/home/dsyer/.m2/repository/javax/ws/rs/jsr311-api/1.1.1/jsr311-api-1.1.1.jar, file:/home/dsyer/.m2/repository/com/netflix/servo/servo-core/0.7.4/servo-core-0.7.4.jar, file:/home/dsyer/.m2/repository/com/sun/jersey/jersey-core/1.13/jersey-core-1.13.jar, file:/home/dsyer/.m2/repository/com/sun/jersey/jersey-client/1.13/jersey-client-1.13.jar, file:/home/dsyer/.m2/repository/com/sun/jersey/contribs/jersey-apache-client4/1.11/jersey-apache-client4-1.11.jar, file:/home/dsyer/.m2/repository/org/apache/httpcomponents/httpclient/4.3.6/httpclient-4.3.6.jar, file:/home/dsyer/.m2/repository/org/apache/httpcomponents/httpcore/4.3.3/httpcore-4.3.3.jar, file:/home/dsyer/.m2/repository/commons-codec/commons-codec/1.6/commons-codec-1.6.jar, file:/home/dsyer/.m2/repository/com/netflix/ribbon/ribbon-httpclient/2.0.0/ribbon-httpclient-2.0.0.jar, file:/home/dsyer/.m2/repository/commons-collections/commons-collections/3.2.1/commons-collections-3.2.1.jar, file:/home/dsyer/.m2/repository/com/netflix/netflix-commons/netflix-commons-util/0.1.1/netflix-commons-util-0.1.1.jar, file:/home/dsyer/.m2/repository/com/netflix/governator/governator/1.2.10/governator-1.2.10.jar, file:/home/dsyer/.m2/repository/com/google/inject/guice/3.0/guice-3.0.jar, file:/home/dsyer/.m2/repository/javax/inject/javax.inject/1/javax.inject-1.jar, file:/home/dsyer/.m2/repository/org/springframework/cloud/spring-cloud-starter-ribbon/1.0.1.RELEASE/spring-cloud-starter-ribbon-1.0.1.RELEASE.jar, file:/home/dsyer/.m2/repository/com/netflix/ribbon/ribbon/2.0.0/ribbon-2.0.0.jar, file:/home/dsyer/.m2/repository/com/netflix/ribbon/ribbon-transport/2.0.0/ribbon-transport-2.0.0.jar, file:/home/dsyer/.m2/repository/io/reactivex/rxnetty-contexts/0.4.6/rxnetty-contexts-0.4.6.jar, file:/home/dsyer/.m2/repository/io/reactivex/rxnetty-servo/0.4.6/rxnetty-servo-0.4.6.jar, file:/home/dsyer/.m2/repository/com/netflix/hystrix/hystrix-core/1.4.4/hystrix-core-1.4.4.jar, file:/home/dsyer/.m2/repository/io/reactivex/rxjava/1.0.6/rxjava-1.0.6.jar, file:/home/dsyer/.m2/repository/io/reactivex/rxnetty/0.4.6/rxnetty-0.4.6.jar, file:/home/dsyer/.m2/repository/io/netty/netty-codec-http/4.0.25.Final/netty-codec-http-4.0.25.Final.jar, file:/home/dsyer/.m2/repository/io/netty/netty-codec/4.0.25.Final/netty-codec-4.0.25.Final.jar, file:/home/dsyer/.m2/repository/io/netty/netty-handler/4.0.25.Final/netty-handler-4.0.25.Final.jar, file:/home/dsyer/.m2/repository/io/netty/netty-transport-native-epoll/4.0.25.Final/netty-transport-native-epoll-4.0.25.Final.jar, file:/home/dsyer/.m2/repository/io/netty/netty-common/4.0.25.Final/netty-common-4.0.25.Final.jar, file:/home/dsyer/.m2/repository/io/netty/netty-buffer/4.0.25.Final/netty-buffer-4.0.25.Final.jar, file:/home/dsyer/.m2/repository/io/netty/netty-transport/4.0.25.Final/netty-transport-4.0.25.Final.jar, file:/home/dsyer/.m2/repository/com/netflix/ribbon/ribbon-core/2.0.0/ribbon-core-2.0.0.jar, file:/home/dsyer/.m2/repository/commons-lang/commons-lang/2.6/commons-lang-2.6.jar, file:/home/dsyer/.m2/repository/com/netflix/ribbon/ribbon-loadbalancer/2.0.0/ribbon-loadbalancer-2.0.0.jar, file:/home/dsyer/.m2/repository/com/netflix/netflix-commons/netflix-statistics/0.1.1/netflix-statistics-0.1.1.jar, file:/home/dsyer/.m2/repository/com/netflix/rxjava/rxjava-core/0.20.7/rxjava-core-0.20.7.jar, file:/home/dsyer/.m2/repository/com/netflix/ribbon/ribbon-eureka/2.0.0/ribbon-eureka-2.0.0.jar, file:/home/dsyer/.m2/repository/com/thoughtworks/xstream/xstream/1.4.2/xstream-1.4.2.jar, file:/home/dsyer/.m2/repository/xmlpull/xmlpull/1.1.3.1/xmlpull-1.1.3.1.jar, file:/home/dsyer/.m2/repository/xpp3/xpp3_min/1.1.4c/xpp3_min-1.1.4c.jar, file:/home/dsyer/.m2/repository/org/springframework/boot/spring-boot-starter-aop/1.2.3.RELEASE/spring-boot-starter-aop-1.2.3.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/spring-aop/4.1.6.RELEASE/spring-aop-4.1.6.RELEASE.jar, file:/home/dsyer/.m2/repository/aopalliance/aopalliance/1.0/aopalliance-1.0.jar, file:/home/dsyer/.m2/repository/org/aspectj/aspectjrt/1.8.5/aspectjrt-1.8.5.jar, file:/home/dsyer/.m2/repository/org/aspectj/aspectjweaver/1.8.5/aspectjweaver-1.8.5.jar, file:/home/dsyer/.m2/repository/org/springframework/retry/spring-retry/1.1.2.RELEASE/spring-retry-1.1.2.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/cloud/spring-cloud-config-client/1.0.1.RELEASE/spring-cloud-config-client-1.0.1.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/cloud/spring-cloud-commons/1.0.1.RELEASE/spring-cloud-commons-1.0.1.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/security/spring-security-crypto/3.2.7.RELEASE/spring-security-crypto-3.2.7.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/cloud/spring-cloud-context/1.0.1.RELEASE/spring-cloud-context-1.0.1.RELEASE.jar, file:/home/dsyer/.m2/repository/javax/validation/validation-api/1.1.0.Final/validation-api-1.1.0.Final.jar, file:/home/dsyer/.m2/repository/org/springframework/boot/spring-boot-starter-actuator/1.2.3.RELEASE/spring-boot-starter-actuator-1.2.3.RELEASE.jar, file:/home/dsyer/.m2/repository/org/springframework/boot/spring-boot-actuator/1.2.3.RELEASE/spring-boot-actuator-1.2.3.RELEASE.jar]
```

(You can see that if you set `logging.level.org.springframework.retry: DEBUG` in `bootstrap.yml`.)

**Comment by talksinmath** [comment]
Sorry - but I feel a little stupid right now. Seems like I had a problem in my local .m2 repo and was getting stale dependencies...

Thanks again - much appreciated.

**Pull request closed without merging**